modules/core/lib/Auth/Process/PairwiseID.php

```diff
@@ -0,0 +1,134 @@
+
+namespace SimpleSAML\Module\core\Auth\Process;
+
+use Exception;
+use SAML2\Constants;
+use SAML2\XML\saml\NameID;
+use SimpleSAML\Assert\Assert;
+use SimpleSAML\Auth;
+use SimpleSAML\Utils;
+
+/**
+ * Filter to generate the Pairwise ID attribute.
+ *
+ * See: http://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/csprd01/saml-subject-id-attr-v1.0-csprd01.html
+ *
+ * By default, this filter will generate the ID based on the UserID of the current user.
+ * This is generated from the attribute configured in 'identifyingAttribute' in the
+ * authproc-configuration.
+ *
+ * Example - generate from attribute:
+ * <code>
+ * 'authproc' => [
+ *   50 => [
+ *       'core:PairwiseID',
+ *       'identifyingAttribute' => 'uid',
+ *       'scope' => 'example.org',
+ *   ]
+ * ]
+ * </code>
+ *
+ * @package SimpleSAMLphp
+ */
+class PairwiseID extends Auth\ProcessingFilter
+{
+    /**
+     * The attribute we should generate the pairwise id from.
+     *
+     * @var string
+     */
+    private string $identifyingAttribute;
+
+    /**
+     * The scope to use for this attribute.
+     *
+     * @var string
+     */
+    private string $scope;
+
+    /**
+     * @var \SimpleSAML\Utils\Config|string
+     * @psalm-var \SimpleSAML\Utils\Config|class-string
+     */
+    protected $configUtils = Utils\Config::class;
+
+
+    /**
+     * Initialize this filter.
+     *
+     * @param array &$config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct(array &$config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+
+        Assert::keyExists($config, 'identifyingAttribute', "Missing mandatory 'identifyingAttribute' config setting.");
+        Assert::keyExists($config, 'scope', "Missing mandatory 'scope' config setting.");
+        Assert::stringNotEmpty($config['identifyingAttribute']);
+        Assert::regex(
+            $config['scope'],
+            '/^[a-zA-Z0-9.-]+$/i',
+            'PairwiseID: \'scope\' contains illegal characters.'
+        );
+
+        $this->identifyingAttribute = $config['identifyingAttribute'];
+        $this->scope = $config['scope'];
+    }
+
+
+    /**
+     * Apply filter to add the Pairwise ID.
+     *
+     * @param array &$state  The current state.
+     */
+    public function process(array &$state): void
+    {
+        Assert::keyExists($state, 'Attributes');
+        Assert::keyExists(
+            $state['Attributes'],
+            $this->identifyingAttribute,
+            sprintf(
+                "core:PairwiseID: Missing attribute '%s', which is needed to generate the Pairwise ID.",
+                $this->identifyingAttribute
+            )
+        );
+
+        $userID = $state['Attributes'][$this->identifyingAttribute][0];
+        Assert::regex(
+            $userID,
+            '/^[a-zA-Z0-9=-]+$/i',
+            'PairwiseID: \'identifyingAttribute\' contains illegal characters.'
```

Comment:
I don't disagree with you, however;

3.4.1                 Syntax and Handling

The requirements for this Attribute are identical to those described in Section 3.3.1.

http://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/csprd01/saml-subject-id-attr-v1.0-csprd01.html#_Toc497819495

Reply:
I don't think so. "this Attribute" here means (logically) the output/resulting attribute. Whatever information/input we use to generate it, is not relevant because our generated value always conforms to the syntax. It should e.g. be fine to use a personnel number as the identifingAttribute for which we calculate the hash.

Reply:
I agree it should be fine.. Removed the assertion in favour of a check for non-empty string